design-proposals/shadow-node.md

```diff
@@ -0,0 +1,286 @@
+    - the storage kubevirt takes.
+- Adding virt-controller as a "proxy" agent, we now depend on its availability. This could cause an increase in heartbeat timeout in corner cases like upgrade.
+- To a degree, it is still possible to manipulate pods, by causing VMs to migrate from other nodes.
+
+### API Examples
+```yaml
+apiVersion: <apiversion>
+kind: ShadowNode
+metadata:
+  name: <Equivalent to existing name>
+  # Labels and Annotations only
+spec: {} //Empty, allows further extension
+status: {} //Empty, allows further extension
+```
+
+### Solution option #2: Creating a REST/GRPC communication between virt-handler and virt-controller
+Pass the labels/annotations through a REST/GRPC channel to the virt-controller. REST option is preferred because the server-side is already implemented.
+
+Virt-handler (Server): Instead of directly patching labels/annotation to the node, Virt-handler will maintain a "cache" (NodeInfoMgr) where all the components will "set" their data.
+Virt handler can use its existing [web-service](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/cmd/virt-handler/virt-handler.go#L556) to expose one more endpoint - GetNodeInfo. Once reached - this endpoint will return the node data.
+
+Virt-Controller (Client): Add a new web-service client infra that will:
+1. run as a go routine and poll the client for updates every one minute.This frequency is chosen because it is the highest of the frequencies (=heartbeat).
+2. ask every Virt-handler server separately for the NodeInfo (need to make sure it's possible to address them in a unique manner).
+3. filter non-kubevirt owned annotations/labels.
+4. patch the node if necessary.
+
+The heartbeat annotation and virt-controller's [Node-controller](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/pkg/virt-controller/watch/node.go) can be removed in favor for the web-service timeout mechanism (if received timeout - set node as unreachable).
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- Using existing web-server infra on the server side.
+- security:
+  - the web-service is secure using the cert-manager internal API.
+  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
+- Scalability:
+  - Reduces #API requests (i.e. node patches) to a minimum.
+
+Cons:
+- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.
+- Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.
```

Comment:
minimizing the window removed quite a bit of experienced issues around scheduling. Sacrificing this for an not preferrable approach does not sound like a good idea.

Reply:
Hey @rmohr - I was not aware of this fixing user-experience issues. This would be a good reason to keep it. Can you specify which scenarios were involved?

I wonder how often does these scenarios occur (all these add the `schedulable: false` on teardown):
- virt-handler crashing
- heartbeat `stopCh` channel closing

The third scenario where virt-handler sets the label to false as soon as possible (until heartbeat sets it to "ready" later) - can maybe a good culprit. I would need to measure how much time it takes to be more certain.. 

But in any case I think that at the very least we can agree that setting the label to false on virt-handler is best-effort, and racy by design, so we need a good cause to keep it.

Reply:
Maybe I misunderstood your pint. I thought you were considering changing that we set schedulable to false fast when shutting down (not crash). For crashes virt-controller already sets it to false.

Reply:
my bad - I didn't distinguish between crash and shut-down.

What I mean is - instead of patching the node with the label like before, for crash scenarios we can for example utilize device-manager:
- when virt-handler crashes it severs the device-manager-to-kubelet grpc connection, so the devices change to 0, like so:
```
$ cluster-up/kubectl.sh get node node01 -ojson | jq .status.allocatable | grep kubevirt.io
selecting podman as container runtime
  "devices.kubevirt.io/kvm": "0",
  "devices.kubevirt.io/tun": "0",
  "devices.kubevirt.io/vhost-net": "0",
```
We can rely on this behavior and set the `schedulable: false` label according to `devices.kubevirt.io/kvm`'s quantity. This can be done by virt-controller's node-watcher as well..
- When virt-handler starts and registers to kubelet, it will initially set devices.kubevirt.io/kvm`'s quantity to 0 as well until virt-handler is ready (the first `schedulable: true` label was set at the first heartbeat cycle so that would be the time when it is set to the usual "1k" quantity) so that VMI won't get scheduled to that node before that.

Reply:
Another way is to add a liveness probe to virt-handler, and when the virt-handler Demonset is not "ready" - then virt-controller will find out which virt-handler is missing and set that `schedulable: false` label